Add support for width-aware noon and midnight options for DayPeriod

nordzilla opened the pull request:
Fixes #435 

This patch adds functionality and tests for width-aware `DayPeriod` patterns for `noon` and `midnight`.

https://unicode.org/reports/tr35/tr35-dates.html#dfst-period

According to the spec, when formatting with this pattern:

* The `noon` options are used from `12:00:00` before `12:01:00`
* The `midnight` options are used from `00:00:00` before `00:01:00`
* If the locale does not support `noon` and/or `midnight` the appropriate `AmPm` option will be used as a fallback.
* All other times will continue to use `AmPm` as expected.


There is no public functionality in icu4x to create a `DateTimeFormat` with a custom pattern. The function that handles formatting with a pattern, `write_pattern()`, lives in the `format` module, which is private. In order for tests to have access to the `write_pattern()` function, all testing must live within the `format` module itself, so this patch restructures the previous `format.rs` module into `format/mod.rs` and adds a `format/tests` directory where the relevant testing lives.

nordzilla commented on components/datetime/src/format/mod.rs:
```diff
@@ -143,42 +146,18 @@ where
-            }
-        }
-    }
-}
```
This test moved to the new file in `tests/mod.rs`

sffc replied:
Is this a conventional place to save tests? @zbraniecki @Manishearth

sffc replied:
Why not */components/datetime/tests/dayperiod.rs* ?

sffc replied:
Given that you are writing these tests as parameterized tests with fixtures, I definitely think that they belong as an integration test, not hidden in a test directory in the library.

(sorry for the triple-comment)

nordzilla replied:
I would have liked for it to be an integration test, but as I noted in the description of the PR, there is no public functionality to format dates with a custom pattern, so it has to live within the  private`format` module. This seemed, to me, to be the best compromise, but I'm open to other suggestions.

sffc replied:
Hmm.  This is an interesting question.  You could, in principle, do this as an integration test with a custom data provider that feeds in your custom pattern strings.  That should actually be pretty easy; see [StructProvider](https://unicode-org.github.io/icu4x-docs/doc/icu_provider/struct_provider/struct.StructProvider.html).  Let's discuss.

Manishearth replied:
For unit tests it's fine to create a `tests` module. But if this can be an integration test, that may be better.

nordzilla replied:
@sffc So I looked into this a bit more this morning, specifically into how I might use `StructProvider`.

Even if I used `StructProvider` to get a custom payload, it seems that both [`write_pattern()`](https://github.com/unicode-org/icu4x/blob/273ec285057bad902c6a08880df4c07a188a7968/components/datetime/src/format.rs#L91) and [`DateTimeFormat`](https://github.com/unicode-org/icu4x/blob/273ec285057bad902c6a08880df4c07a188a7968/components/datetime/src/lib.rs#L112) must take a [`DatesV1`](https://github.com/unicode-org/icu4x/blob/273ec285057bad902c6a08880df4c07a188a7968/components/provider/src/structs/dates.rs#L14) struct specifically as their data. 

Unless there is another way to format the datetime using the custom provider, I'm not seeing any viable options to format with a custom pattern publicly. That being said, I'm still trying to build an overall familiarity with the code base, so I certainly could have missed a key insight here.

sffc replied:
Right.  Create a DatesV1, either using `Default::default()` or by loading a stock locale from test data, then modify a field like `patterns.time.full` with your desired pattern.

I like the idea of data-driven tests being able to overwrite locale data for that testing.  @zbraniecki  thoughts?

nordzilla replied:
Okay, so I've now moved this test back to the file where it existed previously, and I moved the new tests to the `components/datetime/tests` folder to be integration tests. I really like this a lot more. 

But it feels somewhat hacky to me to grab the CLDR data for the locale, and then just overwrite the `long` patterns for each relevant case in my test data. Is this really intended functionality? I had previously overlooked that every member in `DatesV1` is public. 

I had a conversation with @zbraniecki about how custom patterns are often misused in practice. My understanding is that this is why we don't really have a public functionality to directly do something like `DateTimeFormat::try_format(&date_time, &pattern)`, and that we prefer to have people create a `DateTimeFormat` using the options bags. 

It just seems strange to me to have this preference to funnel people through pre-existing options, but still allow "backdoors" for custom patterns by letting people directly mutate the data. I'm almost of the opinion that, if the functionality exists anyway, why not provide something like`DateTimeFormat::try_format(&date_time, &pattern)`?

I suppose I just want to gain clarity around the design decisions behind _not_ having a custom format with pattern function, and for making all of the `DatesV1` members public, since I wasn't involved in the project at the time when these were conceived or implemented.

In the interest of not having duplicated conversations, if there are links to relevant pre-existing conversations that would be helpful to me.

sffc replied:
> In the interest of not having duplicated conversations, if there are links to relevant pre-existing conversations that would be helpful to me.

We've had conversations about this before, but I think they weren't written down.

The fields in all the data structs are public by design. The data struct is designed to be the place where you can "overlay" stock data with custom overrides.

That being said, I agree that in this particular case, it would probably be cleaner to have a low level pattern API. We can revisit this test when we have such API.

nordzilla commented on components/datetime/src/provider.rs:
```diff
@@ -180,22 +181,26 @@ impl DateTimeDates for structs::dates::gregory::DatesV1 {
+                } else {
+                    &symbols.pm
+                }
+            }
```
It is still unclear to me if this should handle the `24:00` cases. As we discussed, `24:00` doesn't exist in ISO time:
https://en.wikipedia.org/wiki/ISO_8601#Times

But this case is still included in the `DayPeriod` spec:
https://unicode.org/reports/tr35/tr35-dates.html#Day_Period_Rules

sffc replied:
`u8::from(hour)` should not be able to produce 24.  I hope to fix this as part of #355

nordzilla replied:
Sounds good. I will remove this logic.

zbraniecki commented on components/provider_cldr/src/transform/dates.rs:
```diff
@@ -353,6 +355,8 @@ pub(self) mod cldr_json {
         ["am", am, Cow<'static, str>],
         ["pm", pm, Cow<'static, str>],
+        ["noon", noon, Option<Cow<'static, str>>],
+        ["midnight", midnight, Option<Cow<'static, str>>],
```
can you use the same trick we use in Plurals to omit serialization of optional fields here: https://github.com/unicode-org/icu4x/blob/master/components/provider/src/structs/plurals.rs#L36-L37

This will make the resources smaller.

zbraniecki replied:
In fact, it would be really cool to skip widths that fallback onto other widths that are exactly the same. For example for French it seems like we could have just one width stored since all the specializations are the same.
@sffc - wdyt?

(not a blocker, may be worth filing a separate issue to do the same for DateTime)

sffc replied:
> In fact, it would be really cool to skip widths that fallback onto other widths that are exactly the same. For example for French it seems like we could have just one width stored since all the specializations are the same.

This is in scope of #257.

nordzilla replied:
Sounds reasonable to do this in a separate PR, then.

zbraniecki reviewed (approved):
lgtm!

sffc reviewed (changes requested):
(setting review bit)

sffc commented on components/provider/src/structs/dates.rs:
```diff
@@ -99,7 +99,13 @@ pub mod gregory {
+        day_periods,
+        am: Cow<'static, str>,
+        pm: Cow<'static, str>,
+        noon: Option<Cow<'static, str>>,
```
Copy the `serialize_none` boilerplate from one of the other provider structs

zbraniecki commented on components/datetime/src/provider.rs:
```diff
@@ -180,22 +181,25 @@ impl DateTimeDates for structs::dates::gregory::DatesV1 {
-            &symbols.am
-        } else {
-            &symbols.pm
+        match (day_period, u8::from(hour), u8::from(minute)) {
```
Pure brainstorming: I'm wondering if we can do something to improve this matching. In my naive understanding, the term like `noon` or `midnight` is used in relation to the level of detail in the time presented.

If we're talking about hour+minute, then midnight is at 00:00, but if we are talking about just hour, then midnight is also 00:05.
Similarly if I'm formatting `12:00:05` then it is not `noon`, but if I'm formatting `12:00` then it is even if it is 5 seconds later.

Could we pass the information on the time tokens to be displayed to this function and use that to evaluate it?
Is it worth it?

nordzilla replied:
So what you're saying is we may want to achieve:

**(pseudo code)**
```rust
DateTime::try_from("2021-01-15T00:00:00")?.try_format("HH b")?       == "00 midnight"
DateTime::try_from("2021-01-15T00:00:00")?.try_format("HH:mm b")?    == "00:00 midnight"
DateTime::try_from("2021-01-15T00:00:00")?.try_format("HH:mm:ss b")? == "00:00:00 midnight"

DateTime::try_from("2021-01-15T00:00:01")?.try_format("HH b")?       == "00 midnight"
DateTime::try_from("2021-01-15T00:00:01")?.try_format("HH:mm b")?    == "00:00 midnight"
DateTime::try_from("2021-01-15T00:00:01")?.try_format("HH:mm:ss b")? == "00:00:01 AM"

DateTime::try_from("2021-01-15T00:01:00")?.try_format("HH b")?       == "00 midnight"
DateTime::try_from("2021-01-15T00:01:00")?.try_format("HH:mm b")?    == "00:01 AM"
DateTime::try_from("2021-01-15T00:01:00")?.try_format("HH:mm:ss b")? == "00:01:00 AM"

DateTime::try_from("2021-01-15T01:00:00")?.try_format("HH b")?       == "01 AM"
DateTime::try_from("2021-01-15T01:00:00")?.try_format("HH:mm b")?    == "01:00 AM"
DateTime::try_from("2021-01-15T01:00:00")?.try_format("HH:mm:ss b")? == "01:00:00 AM
```
Is that correct?

zbraniecki replied:
I don't think day periods are very useful in h23/h24, but yeah. `12 midnight` and `12:00 midnight` but `12:00:01 am`.

sffc replied:
Did we decide to do this in this PR or follow up later?

nordzilla replied:
My understanding, based on a conversation in Slack, is that we still have not gained a clarity or consensus on what improvements we want to make here. 

@mihnita suggested that we should move this discussion to its own issue. 

@zbraniecki I think it's reasonable to land this now to avoid bitrot while we figure out what improvements ICU4X can make in this area.

I'm happy to create the issue for this.

mihnita replied:
+1 to land without noon / midnight

nordzilla replied:
I think that is something that needs to be discussed then, even local to this GitHub issue, since that would mean not landing this PR at all.

mihnita replied:
OK... reconsidering (discussed in meeting)

This is about patterns.
They are "supposed" to be "dumb", as in "do exactly what I'm telling you, even if bad i18n"
The real i18n issue will be in resolving the skeletons to patterns.

So OK to land as is (I will approve)

zbraniecki reviewed (approved):
lgtm!

nordzilla commented:
I just want to clarify everything that we have discussed formally in a comment.

So the goal is to add two new sets of functionality to this PR:

1) Modify the `DataProvider` to generate special patterns, as @sffc noted (pseudo-syntax)
```c++
if (m == 0) {
  if (h == 0) {
    "'midnight'"
  } else if (h == 12) {
    "'noon'"
  } else {
    "h a"
  }
} else {
  "h:mm a"
}
```

2) As `DateTimeFormat` is constructed, when tokenizing the pattern, determine which dayperiod fields are needed based on how specific the display is. 

Is this correct?  

I think that number 2) makes sense, I am still a bit unclear still about number 1) because the `DataProvider` isn't necessarily aware of any specific datetime. What does `m == 0` mean in the context of only the `DataProvider`?

It's the `DateTimeFormat` that puts this all together. I would think all of this logic should live in `DateTimeFormat`.

sffc commented:
The pseudo-syntax I posted above is something we should do later in #380.  For now we just need to do what you posted in https://github.com/unicode-org/icu4x/pull/444#discussion_r558681578, which is a bit simpler.

nordzilla commented:
As a first pass to achieve the functionality described in https://github.com/unicode-org/icu4x/pull/444#discussion_r558681578, I have naively added the capability for a `Pattern` to retrieve the maximum time granularity contained inside the pattern.

Currently, the maximum granularity is computed every time the pattern is written. This is obviously going to be slower than computing the value once and caching the result, but it was one of the simplest approaches that touched the least amount of existing code in order to get new tests written for the functionality. 

I am curious to see if and how much this regresses the benchmarks. I suspect that I will likely add one more commit to compute and cache this value.

zbraniecki commented on components/datetime/src/format/mod.rs:
```diff
@@ -95,6 +99,17 @@ pub fn write_pattern<T>(
+        Some(TimeGranularity::Seconds) => {
+            u8::from(date_time.minute()) + u8::from(date_time.second()) == 0
+        }
+    };
```
Since `b` is a rare field, I'd suggest to calculate all of that lazily only if you encounter `b`. You could have `get_time_granularity_for_pattern(pattern) -> Option<Granularity>` function for that.

zbraniecki commented on components/datetime/src/format/mod.rs:
```diff
@@ -1,11 +1,15 @@
 // This file is part of ICU4X. For terms of use, please see the file
```
Why did you turn `format.rs` into `format/mod.rs`?

nordzilla replied:
Oops, this was leftover from before we were using `StructProvider` to let these tests be integration tests. It looks like I missed this change when I was converting everything back. I will return it to how it was.

zbraniecki commented on components/datetime/src/pattern/mod.rs:
```diff
@@ -16,6 +16,28 @@ pub enum PatternItem {
+            _ => None,
+        }
+    }
+}
```
I don't think having it here has any performance/size implications, but I think from the architectural standpoint that method located here is misplaced.
It communicates to me that there's a need to calculate `time_granularity` for `PatternItem` elements, while in fact, there's only one scenario (so far?) where we need it and it's an outlier.

For cases where a bunch of code is used only for a specialized scenario, it is a good idea to:
* Keep that code isolated in some helper function/module
* Link it only from the direct code that needs it. In this case, call it only from `b` match case. This will allow people later to easily grep on why the function exists, where it is used and either find a callee or a caller when needed.
* Document the shit out of that special function (`get_time_granularity_for_pattern`) so that people understand what it does and why when they reach that code

nordzilla replied:
Thank you for your feedback (to this and related comments). I'm totally happy to, and was planning to reorganize a lot of this. As I said in https://github.com/unicode-org/icu4x/pull/444#issuecomment-765809418, I just wanted to do a first pass from a path of least resistance to ensure that the new functionality was tested (I love having tests to back me up).

Your comments give me good direction for how to rearrange all of this.

zbraniecki commented on components/datetime/src/pattern/mod.rs:
```diff
@@ -62,6 +84,15 @@ impl Pattern {
+    /// e.g. The maximum granularity for `E, dd/MM/y` is `None`.
+    pub fn most_granular_time(&self) -> Option<TimeGranularity> {
+        self.0.iter().flat_map(PatternItem::time_granularity).max()
+    }
```
Same comment as above. I wouldn't put such specialized methods on `PatternItem`. Maybe even add a `relative_time` module with helper functions for calculations around it.

It's purely a recommendation for code readability.

zbraniecki commented on components/datetime/src/provider.rs:
```diff
@@ -180,22 +181,25 @@ impl DateTimeDates for structs::dates::gregory::DatesV1 {
+                } else {
+                    &symbols.pm
+                }
+            }
```
optional nit; I like using match guards for cases like this:

```
(_, hour, _) if hour < 12 => &symbols.am,
(_, hour, _) => &symbols.pm
```

but that's totally up to you.

sffc commented on components/datetime/src/format.rs:
```diff
@@ -95,6 +99,48 @@ fn get_day_of_week(year: usize, month: date::Month, day: date::Day) -> date::Wee
+    date_time: &T,
+) -> bool {
+    matches!(day_period, DayPeriod::NoonMidnight)
+        && match pattern.0.iter().flat_map(time_granularity).max() {
```
You're scanning the pattern each time?

nordzilla replied:
I've updated this to be a cached, lazily-initialized member of `Pattern`.

sffc commented on components/datetime/src/format.rs:
```diff
@@ -95,6 +99,48 @@ fn get_day_of_week(year: usize, month: date::Month, day: date::Day) -> date::Wee
+/// if the most granular time being displayed lands exactly on the hour. For example, the time
+/// 12:00:00 is always noon-compatible. The time 12:05:15 is noon-compatible only if the
+/// pattern does not contain the minutes or seconds.
+fn is_noon_midnight_compatible<T: DateTimeType>(
```
How about `is_top_of_hour`?

nordzilla replied:
I like it. Done.

sffc commented on components/datetime/src/format.rs:
```diff
@@ -95,6 +99,48 @@ fn get_day_of_week(year: usize, month: date::Month, day: date::Day) -> date::Wee
+    date_time: &T,
+) -> bool {
+    matches!(day_period, DayPeriod::NoonMidnight)
+        && match pattern.0.iter().flat_map(time_granularity).max() {
```
You may be able to save metadata in additional fields of `pattern` as the pattern is being parsed.

sffc reviewed (approved):
LGTM!

nordzilla commented:
All right @sffc I rebased on top of master after you landed #451.

This should be ready for one final look (no changes since last time), and then able to merge with no conflicts.

nordzilla commented:
Uh-oh, spoke too soon. Looks like the CI is having an issue with the optional serde serialization, even though it works for me locally.

It probably has something to do with you making serde optional. I will look into this.

nordzilla commented:
@sffc all fixed. The CI failure was due to me not handling optional serde in the macro rewrite. Should be good to go now.

sffc commented on components/datetime/src/pattern/mod.rs:
```diff
@@ -48,25 +48,67 @@ impl<'p> From<String> for PatternItem {
+impl From<Vec<PatternItem>> for Pattern {
+    fn from(items: Vec<PatternItem>) -> Self {
+        Self {
+            time_granularity: items.iter().flat_map(get_time_granularity).max(),
```
Optional: `filter_map` instead of `flat_map`.  I spent several minutes trying to figure out why `flat_map` worked here; `filter_map` would be more readable and perhaps a bit faster since it's less general.

nordzilla replied:
Totally, happy to use `filter_map` here. Done.

zbraniecki reviewed (approved):
lgtm!
